### PR TITLE
Add CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by felixbarny**

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+### Workflow
+
+All feature development and most bug fixes hit the master branch first.
+Pull requests should be reviewed by someone with commit access.
+Once approved, the author of the pull request,
+or reviewer if the author does not have commit access,
+should "Squash and merge".
+
+
```
Comment (watson) [inline CONTRIBUTING.md]: Nit: Remove double line break

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    think about if we can add the necessary information to make diagnosing similar issues easier.
+  * Try to make exception messages as descriptive as possible.
+    When this particular exception is thrown,
+    what is the user supposed to do to fix the problem?
+  * Sometimes the best debug logs are just not enough to diagnose a problem.
+    The agent code should be easily debuggable in the user's IDE.
+* Performance is a feature
+  * Find the right balance of low-level and high-level code
+  * We are especially proud of the low allocation rate of the agent.
+    This does not only the average overhead look good,
```
Comment (axw) [inline CONTRIBUTING.md]: I think you're missing a word here ("make"?)

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+### Coding guidelines
+
+There are a few guidelines or goals for the development of the agent.
+Some of these goals may be contrary at times,
+like performance and readability.
+We should try to find the best compromise and balance for each specific case.
+
+Not all of these guidelines are perfectly put into practice and there is always room for improvement.
+But for each change, we should think about whether they bring us closer to or further away from those goals.
```
Comment (watson) [inline CONTRIBUTING.md]: Nit: Add line break after comma

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    Improvements should be verified by the benchmarks.
+    To get more insight about potential areas of improvements,
+    enable Flight Recording during the benchmarks and analyze the results in Java Mission Control.
+* Easy to install for users
+  * Installing the agent should be as easy as possible.
+    The fewer installation steps and required configuration options the user has to deal with,
+    the better.
+* Testability/Test Pyramid
+  * Most tests should be simple and easily debuggable JUnit tests.
+    In fact, the main workflow should be to tests alongside the production code.
```
Comment (axw) [inline CONTRIBUTING.md]: "should be to ___ tests"?
or "should be to test"?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    You should rarely have to build the agent and attach it to a test application in order to verify your new code works as expected.
+    This makes turnarounds much faster and usually also increases the test coverage.
+    Speaking of which,
+    use code coverage tools to check that the most important branches are covered by unit tests.
+    The coverage of new code should not be lower than the current total coverage of the project.
```
Comment (axw) [inline CONTRIBUTING.md]: I'd temper that with "typically", and say that coverage is not a goal in and of itself and to use judgement.
(highly subjective area, so feel free to ignore me - I'm not maintaining this repo :))

File: CONTRIBUTING.md (hunk 4 above)
Comment (felixbarny) [reply CONTRIBUTING.md]: to write tests

File: CONTRIBUTING.md (hunk 5 above)
Comment (felixbarny) [reply CONTRIBUTING.md]: Totally agree. I recently did a change which made the coverage tool not recognize that a certain file is tested at all.

**Pull request merged**